scala/private/rule_impls.bzl

```diff
@@ -0,0 +1,756 @@
+    if getattr(ctx.attr, "main_class", ""):
+        args.extend(["--main_class", ctx.attr.main_class])
+    args.extend(["--output", ctx.outputs.deploy_jar.path])
+    ctx.actions.run(
+        inputs=jars_list,
+        outputs=[ctx.outputs.deploy_jar],
+        executable=ctx.executable._singlejar,
+        mnemonic="ScalaDeployJar",
+        progress_message="scala deployable %s" % ctx.label,
+        arguments=args)
+
+def _path_is_absolute(path):
+    # Returns true for absolute path in Linux/Mac (i.e., '/') or Windows (i.e.,
+    # 'X:\' or 'X:/' where 'X' is a letter), false otherwise.
+    if len(path) >= 1 and path[0] == "/":
+        return True
+    if len(path) >= 3 \
+            and path[0].isalpha() \
+            and path[1] == ":" \
+            and (path[2] == "/" or path[2] == "\\"):
+        return True
+
+    return False
+
+def _runfiles_root(ctx):
+    return "${TEST_SRCDIR}/%s" % ctx.workspace_name
+
+def _write_java_wrapper(ctx, args="", wrapper_preamble=""):
+    """This creates a wrapper that sets up the correct path
+       to stand in for the java command."""
+
+    java_path = str(ctx.attr._java_runtime[java_common.JavaRuntimeInfo].java_executable_runfiles_path)
+    if _path_is_absolute(java_path):
+      javabin = java_path
+    else:
+      runfiles_root = _runfiles_root(ctx)
+      javabin = "%s/%s" % (runfiles_root, java_path)
+
+
+    exec_str = ""
+    if wrapper_preamble == "":
+      exec_str = "exec "
+
+    wrapper = ctx.actions.declare_file(ctx.label.name + "_wrapper.sh")
+    ctx.actions.write(
+        output = wrapper,
+        content = """#!/bin/bash
+{preamble}
+
+{exec_str}{javabin} "$@" {args}
+""".format(
+            preamble=wrapper_preamble,
+            exec_str=exec_str,
+            javabin=javabin,
+            args=args,
+        ),
+        is_executable = True
+    )
+    return wrapper
+
+def _write_executable(ctx, rjars, main_class, jvm_flags, wrapper):
+    template = ctx.attr._java_stub_template.files.to_list()[0]
+    # RUNPATH is defined here:
+    # https://github.com/bazelbuild/bazel/blob/0.4.5/src/main/java/com/google/devtools/build/lib/bazel/rules/java/java_stub_template.txt#L227
+    classpath = ":".join(["${RUNPATH}%s" % (j.short_path) for j in rjars.to_list()])
+    jvm_flags = " ".join([ctx.expand_location(f, ctx.attr.data) for f in jvm_flags])
+    ctx.actions.expand_template(
+        template = template,
+        output = ctx.outputs.executable,
+        substitutions = {
+            "%classpath%": classpath,
+            "%java_start_class%": main_class,
+            "%javabin%": "JAVABIN=%s/%s" % (_runfiles_root(ctx), wrapper.short_path),
+            "%jvm_flags%": jvm_flags,
+            "%needs_runfiles%": "",
+            "%runfiles_manifest_only%": "",
+            "%set_jacoco_metadata%": "",
+            "%workspace_prefix%": ctx.workspace_name + "/",
+        },
+        is_executable = True,
+    )
+
+def dep_target_contains_ijar(dep_target):
```

Comment:
why is this public?

Reply:
I copied it as it was from the original code. It doesn't seem to be used anywhere, so I suppose I could remote it entirely.

Reply:
fine by me.